src/Common/ExponentiallySmoothedCounter.h

```diff
@@ -109,4 +114,375 @@ struct ExponentiallySmoothedAverage
+    }
+
+    /// Merge two counters. It is done by moving to the same point of reference and summing the values.
+    /// First counter will be 'main' one, and second will be 'additional' one.
+    /// So the values of first were at the end, and second - at the beginning of the sequence.
+    static ExponentiallySmoothedAlpha merge(const ExponentiallySmoothedAlpha & a, const ExponentiallySmoothedAlpha & b, double alpha)
+    {
+        return ExponentiallySmoothedAlpha(a.remap(a.count + b.count, alpha).value + b.value, a.count + b.count,
+            a.count > 0 ? a.first_value : b.first_value);
+    }
+
+    void merge(const ExponentiallySmoothedAlpha & other, double alpha)
+    {
+        *this = merge(*this, other, alpha);
+    }
+
+    /// Add new value.
+    void add(double new_value, double alpha)
+    {
+        merge(ExponentiallySmoothedAlpha(new_value * alpha, 1, new_value), alpha);
+    }
+
+    /// Get current value.
+    double get(double alpha) const
+    {
+        return value + first_value * scale(count, alpha);
+    }
+
+    /// Get value at the given count.
+    double get(unsigned long long int current_count, double alpha) const
+    {
+        return remap(current_count, alpha).get(alpha);
+    }
+
+    /// Compare two counters (by moving to the same point of reference and comparing sums).
+    /// You can store the counters in container and sort it without changing the stored values over time.
+    bool less(const ExponentiallySmoothedAlpha & other, double alpha) const
+    {
+        unsigned long long int max_count = std::max(count, other.count);
+        return get(max_count, alpha) < other.get(max_count, alpha);
+    }
+};
+
+struct ExponentiallySmoothedAlphaWithTime
+{
+    /// The sum. It contains the last value and all previous values scaled accordingly to the difference of their time to the reference time.
+    /// Older values are summed with exponentially smaller coefficients.
+
+    double value = 0;
+
+    /// Current timestamp. Using in calculating exponential smoothing.
+
+    unsigned long long int timestamp = 0;
+
+    /// first applied value. Using to avoid multiplying first value on alpha. 
+
+    struct {
+        double value = 0;
+        unsigned long long int timestamp = 0;
+        bool was = false;
+    } first_value;
+
+    ExponentiallySmoothedAlphaWithTime() = default;
+
+    template<typename first_value_type>
+    ExponentiallySmoothedAlphaWithTime(double current_value, unsigned long long int current_time, first_value_type first_value_)
+        : value(current_value), timestamp(current_time), first_value(first_value_)
+    {
+    }
+
+    ExponentiallySmoothedAlphaWithTime(double current_value, unsigned long long int current_time, double first_value_, unsigned long long int first_timestamp_)
+        : value(current_value), timestamp(current_time)
+    {
+        first_value.value = first_value_;
```

Comment:
Maybe you can use this syntax to initialize `first_value` in the initialization list (l. 233):

```cpp
[...], first_value {.value = first_value_, .timestamp = first_timestamp_, was = true}
```